pkg/utils/rbac/access_review.go

```diff
@@ -0,0 +1,142 @@
+	metaObject, err := meta.Accessor(object)
+	if err != nil {
+		return false, nil
+	}
+	// For convenience we still allow association between objects in a same namespace
+	if sourceNamespace == metaObject.GetNamespace() {
+		return true, nil
+	}
+
+	if len(serviceAccount) == 0 {
+		serviceAccount = "default"
+	}
+
+	allErrs := field.ErrorList{}
+	// This validation could be done in other places but it is important to be sure that it is done before any access review.
+	for _, msg := range validation.IsDNS1123Subdomain(serviceAccount) {
+		allErrs = append(allErrs, &field.Error{Type: field.ErrorTypeInvalid, Field: "serviceAccount", BadValue: serviceAccount, Detail: msg})
+	}
+	if len(allErrs) > 0 {
+		return false, allErrs.ToAggregate()
+	}
+
+	kind := object.GetObjectKind().GroupVersionKind().Kind
+	plural, err := toPlural(kind)
+	if err != nil {
+		return false, nil
+	}
+
+	sar := &authorizationapi.SubjectAccessReview{
+		Spec: authorizationapi.SubjectAccessReviewSpec{
+			ResourceAttributes: &authorizationapi.ResourceAttributes{
+				Namespace: metaObject.GetNamespace(),
+				Verb:      "get",
+				Resource:  plural,
+				Group:     strings.ToLower(object.GetObjectKind().GroupVersionKind().Group),
+				Version:   strings.ToLower(object.GetObjectKind().GroupVersionKind().Version),
+				Name:      metaObject.GetName(),
+			},
+			User: ServiceAccountUsernamePrefix + sourceNamespace + ":" + serviceAccount,
+		},
+	}
+
+	sar, err = s.client.AuthorizationV1().SubjectAccessReviews().Create(sar)
+	if err != nil {
+		return false, err
+	}
+	log.V(1).Info(
+		"Access review", "result",
+		sar.Status, "serviceAccount", serviceAccount,
+		"sourceNamespace", sourceNamespace,
+		"remoteKind", kind,
+		"remoteNamespace", metaObject.GetNamespace(),
+		"remoteName", metaObject.GetName(),
+	)
+	if sar.Status.Denied {
+		return false, nil
+	}
+	return sar.Status.Allowed, nil
+}
+
+// Lazy hack to get the plural form
+func toPlural(singular string) (string, error) {
+	switch singular {
+	case "Elasticsearch":
+		return "elasticsearches", nil
+	case "Kibana":
+		return "kibanas", nil
+	case "ApmServer":
+		return "apmservers", nil
+	}
+	return "", fmt.Errorf("unknown singular kind: %s", singular)
+}
+
+type permissiveAccessReviewer struct{}
+
+var _ AccessReviewer = &permissiveAccessReviewer{}
+
+func (s *permissiveAccessReviewer) AccessAllowed(_ string, _ string, _ runtime.Object) (bool, error) {
+	return true, nil
+}
+
+// NextReconciliation returns a reconcile result depending on the implementation of the AccessReviewer.
+// It is mostly used when using the subjectAccessReviewer implementation in which case a next reconcile loop should be
+// triggered later to keep the association in sync with the RBAC roles and bindings.
+// See https://github.com/elastic/cloud-on-k8s/issues/2468#issuecomment-579157063
+func NextReconciliation(accessReviewer AccessReviewer) reconcile.Result {
```

Comment:
Having this in the `rbac` package feels a bit strange to me, it does not keep a clear separation of concerns.
Maybe in `common/association/rbac.go`? We could call it as `association.RequeueRbacCheck()`.
It would force `subjectAccessReviewer ` to become public though. Maybe that's ok? 

(I don't have a strong opinion on this, I'm fine if you prefer keeping the current approach)

Reply:
I was wondering exactly the same. The problem is that this package is supposed to be reused for remote clusters (CCR/CCS) which is not strictly speaking considered (so far) as an association but more as an internal feature of the elasticsearch controller.

Reply:
I have addressed your concern and moved this part of the code to the association package. I will check later if it does fit or not with remote clusters.